virtool/administrators/api.py

```diff
@@ -1,130 +1,156 @@
-from typing import Union
+import asyncio
+from typing import Union, Optional
 
-from aiohttp.web_exceptions import HTTPNoContent, HTTPBadRequest
+from aiohttp.web_exceptions import HTTPForbidden, HTTPBadRequest
 from aiohttp_pydantic import PydanticView
-from aiohttp_pydantic.oas.typing import r201, r200, r204, r404
-
+from aiohttp_pydantic.oas.typing import r201, r200, r404
 from virtool_core.models.roles import AdministratorRole
+from virtool_core.utils import document_enum
 
 from virtool.administrators.oas import (
-    CreateAdministratorRequest,
-    UpdateAdministratorRequest,
+    UpdateAdministratorRoleRequest,
     ListAdministratorResponse,
-    GetAdministratorResponse,
-    CreateAdministratorResponse,
-    UpdateAdministratorResponse,
+    ListRolesResponse,
+    UpdateUserRequest,
+    UserResponse,
 )
 from virtool.api.response import NotFound, json_response
+from virtool.authorization.utils import get_authorization_client_from_req
 from virtool.data.errors import ResourceNotFoundError
 from virtool.data.utils import get_data_from_req
 from virtool.http.policy import policy, AdministratorRoutePolicy
 from virtool.http.routes import Routes
 
 routes = Routes()
 
+AVAILABLE_ROLES = [
+    {"id": role, "name": role.capitalize(), "description": role.__doc__}
+    for role in document_enum(AdministratorRole)
+]
+
 
-@routes.view("/administrators")
+@routes.view("/admin/roles")
 class AdministratorsView(PydanticView):
-    @policy(AdministratorRoutePolicy(AdministratorRole.FULL))
-    async def get(self) -> r200[ListAdministratorResponse]:
+    @policy(AdministratorRoutePolicy(AdministratorRole.BASE))
+    async def get(self) -> r200[ListRolesResponse]:
         """
-        List all administrators.
+        List all available administrator roles.
 
         Status Codes:
             200: Successful operation
         """
-        return json_response(
-            await get_data_from_req(self.request).administrators.find()
-        )
+        return json_response(AVAILABLE_ROLES)
 
-    @policy(AdministratorRoutePolicy(AdministratorRole.FULL))
-    async def post(
-        self, data: CreateAdministratorRequest
-    ) -> Union[r201[CreateAdministratorResponse], r404]:
+
+@routes.view("/admin/users")
+class AdministratorsView(PydanticView):
+    @policy(AdministratorRoutePolicy(AdministratorRole.USERS))
+    async def get(
+        self, administrator: Optional[bool] = None, term: Optional[str] = None
+    ) -> r200[ListAdministratorResponse]:
         """
-        Add a user as an administrator.
+        Get a paginated list of users
 
         Status Codes:
-            201: Successful operation
-            400: User not found
+            200: Successful operation
         """
 
-        try:
-            administrator = await get_data_from_req(self.request).administrators.create(
-                data
+        return json_response(
+            await get_data_from_req(self.request).administrators.find_users(
```

Comment:
Yeah I think your right that it could be parsed via aiohttp. The short version is that when updating `paginate_aggregate` to do the aggregation I was trying to keep the api as consistent with other `paginate` functions as possible. `paginate` expects the `url_query` while `paginate2` expects `page` and `per_page` as `ints`. `paginate` is in use `paginate2` is not.

If you prefer the `paginate2` api I can shift it over to expecting page and per page be handed to it as `ints`. (which imo makes more sense with our current separation of concerns)